File: implementations/go/provekit-ir-symbolic/verifier/load_all_proofs.go
package verifier

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/provekit/ir-symbolic/canonicalizer"
	"github.com/provekit/ir-symbolic/proof_envelope"
)

// LoadAllProofsStage walks every .proof file under projectRoot
// (project root + node_modules/{*,@*/*}/) and builds the unified
// MementoPool downstream stages hash-look-up against.
//
// Implements §3 rule 1 (filename CID matches content) + §3 rule 2
// (member CIDs match envelope identities). Rule 3 (catalog signature
// verify) is deferred — needs a public-key memento walker that v1
// doesn't yet have.
//
// v1.1.0: every protocol-surface hash is BLAKE3-512 with the
// "blake3-512:" tag; every signature is "ed25519:" + base64(sig).
// Unknown algorithm tags fail-loud per the verifier-dispatch contract
// in protocol/specs/2026-04-30-memento-envelope-grammar.md.
type LoadAllProofsStage struct{}

// Tag prefixes permitted on the v1.1.0 protocol surface.
const (
	hashTagPrefix = "blake3-512:"
	sigTagPrefix  = "ed25519:"
)

// Run loads every .proof in projectRoot and returns the unified pool.
func (s *LoadAllProofsStage) Run(projectRoot string) (*MementoPool, error) {
	pool := &MementoPool{
		Mementos:        map[string]map[string]interface{}{},
		BridgesBySymbol: map[string]map[string]interface{}{},
	}
	for _, pp := range enumerateProofFiles(projectRoot) {
		if err := s.loadOne(pp, pool); err != nil {
			pool.LoadErrors = append(pool.LoadErrors, LoadError{ProofPath: pp, Reason: err.Error()})
		}
	}
	return pool, nil
}

// proofFilenameV11RE matches the v1.1.0 filename shape:
//
//	blake3-512:<128 hex>.proof
var proofFilenameV11RE = regexp.MustCompile(`^(blake3-512:[0-9a-f]{128})\.proof$`)

// proofFilenameAnyTagRE matches any self-identifying-tag filename so
// we can reject unknown tags loud (mirrors C++ load_all_proofs.cpp
// re_anytag).
var proofFilenameAnyTagRE = regexp.MustCompile(`^([a-z0-9]+-[0-9]+:[0-9a-f]+)\.proof$`)

// proofFilenameLegacyRE matches the pre-v1.1.0 sha256 32-hex shape
// (no algorithm tag). Treated as a legacy fixture: filename-shape
// passes (no rule-1 hash check), but the member envelopes inside
// will lack "blake3-512:" prefixes and fail loud at the member-tag
// check below. This mirrors the C++ verifier's permissive filename
// behavior for unmarked files while still hard-rejecting v0 hashes.
var proofFilenameLegacyRE = regexp.MustCompile(`^[0-9a-f]+\.proof$`)

func (s *LoadAllProofsStage) loadOne(proofPath string, pool *MementoPool) error {
	bytes, err := os.ReadFile(proofPath)
	if err != nil {
		return fmt.Errorf("read: %w", err)
	}

	// Rule 1: filename CID matches content (trust root).
	// v1.1.0 prefers `blake3-512:<128 hex>.proof`. We also tolerate
	// bare `<hex>.proof` (legacy path; the hash-tag check below catches
	// any v0 mementos inside it). Any OTHER self-identifying tag is
	// rejected loud — mirrors implementations/cpp/provekit/verifier/
	// load_all_proofs.cpp.
	filename := filepath.Base(proofPath)
	if m := proofFilenameV11RE.FindStringSubmatch(filename); m != nil {
		filenameCID := m[1]
		derived := canonicalizer.ComputeCID(bytes)
		if derived != filenameCID {
			return fmt.Errorf("rule 1 (trust root): filename CID %s != content hash %s",
				filenameCID, derived)
		}
	} else if m := proofFilenameAnyTagRE.FindStringSubmatch(filename); m != nil {
		// Self-identifying tag we do not recognize. Reject loud.
		return fmt.Errorf("rule 1 (trust root): unsupported hash tag in filename %q; v1.1.0 requires `blake3-512:`",
			m[1])
	} else if !proofFilenameLegacyRE.MatchString(filename) {
		// Filename doesn't match any known shape; ignore (don't
		// surface as an error — non-.proof file got into the
		// enumerator? shouldn't happen but be lenient).
	}

	dec := proof_envelope.NewCBORDecoder(bytes)
	catalog, err := dec.DecodeCatalog()
	if err != nil {
		return fmt.Errorf("decode: %w", err)
	}
	membersAny, ok := catalog["members"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("catalog has no `members` map")
	}
	for cid, raw := range membersAny {
		// Tag-dispatch: every member CID MUST be self-identifying.
		// v1.1.0 permits only "blake3-512:"; reject anything else loud.
		if !strings.HasPrefix(cid, hashTagPrefix) {
			pool.LoadErrors = append(pool.LoadErrors, LoadError{
				ProofPath: proofPath,
				Reason:    fmt.Sprintf("member %s: unsupported hash tag; v1.1.0 requires `blake3-512:`", cid),
			})
			continue
		}
		envBytes, ok := raw.([]byte)
		if !ok {
			pool.LoadErrors = append(pool.LoadErrors, LoadError{
				ProofPath: proofPath,
				Reason:    fmt.Sprintf("member %s: value is %T, expected bstr", cid, raw),
			})
			continue
		}
		var env map[string]interface{}
		if err := json.Unmarshal(envBytes, &env); err != nil {
			pool.LoadErrors = append(pool.LoadErrors, LoadError{
				ProofPath: proofPath,
				Reason:    fmt.Sprintf("member %s: parse: %s", cid, err),
			})
			continue
		}
		// Tag-dispatch on producerSignature: v1.1.0 permits "ed25519:" only.
		if sigRaw, ok := env["producerSignature"].(string); ok {
			if !strings.HasPrefix(sigRaw, sigTagPrefix) {
				pool.LoadErrors = append(pool.LoadErrors, LoadError{
					ProofPath: proofPath,
					Reason:    fmt.Sprintf("member %s: unsupported signature tag; v1.1.0 requires `ed25519:`", cid),
				})
				continue
			}
		}
		// Rule 2: re-derive member CID from envelope JCS.
		derived, err := computeEnvelopeCID(env)
		if err != nil || derived != cid {
			pool.LoadErrors = append(pool.LoadErrors, LoadError{
				ProofPath: proofPath,
				Reason:    fmt.Sprintf("rule 2: member %s derives to %s", cid, derived),
			})
			continue
		}
		pool.Mementos[cid] = env
		// If it's a bridge envelope, index by sourceSymbol.
		if ev, ok := env["evidence"].(map[string]interface{}); ok {
			if ev["kind"] == "bridge" {
				if body, ok := ev["body"].(map[string]interface{}); ok {
					if sym, ok := body["sourceSymbol"].(string); ok {
						pool.BridgesBySymbol[sym] = env
					}
				}
			}
		}
	}
	return nil
}

// computeEnvelopeCID re-derives the envelope's CID by JCS-encoding the
// envelope minus cid + producerSignature. Per v1.1.0:
//
//	"blake3-512:" + hex(BLAKE3_512(canonical-bytes))
//
// (universal-claim-envelope.md §CID construction).
func computeEnvelopeCID(env map[string]interface{}) (string, error) {
	stripped := make(map[string]interface{}, len(env))
	for k, v := range env {
		if k == "cid" || k == "producerSignature" {
			continue
		}
		stripped[k] = v
	}
	enc := canonicalizer.NewEncoder()
	bytes, err := enc.Encode(stripped)
	if err != nil {
		return "", err
	}
	return canonicalizer.ComputeCID(bytes), nil
}

// enumerateProofFiles walks projectRoot + node_modules/{*,@*/*}/ for
// every *.proof file (one level deep only).
func enumerateProofFiles(projectRoot string) []string {
	var out []string
	pushProofs := func(dir string) {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return
		}
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".proof") {
				out = append(out, filepath.Join(dir, e.Name()))
			}
		}
	}
	pushProofs(projectRoot)

	nodeModules := filepath.Join(projectRoot, "node_modules")
	entries, err := os.ReadDir(nodeModules)
	if err != nil {
		return out
	}
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		if strings.HasPrefix(e.Name(), ".") {
			continue
		}
		entryPath := filepath.Join(nodeModules, e.Name())
		if strings.HasPrefix(e.Name(), "@") {
			scoped, err := os.ReadDir(entryPath)
			if err != nil {
				continue
			}
			for _, sub := range scoped {
				if sub.IsDir() {
					pushProofs(filepath.Join(entryPath, sub.Name()))
				}
			}
		} else {
			pushProofs(entryPath)
		}
	}
	return out
}

File: src/workflow/producers/loadAllProofs.ts
/**
 * load-all-proofs — Stage 0/1 of the bridge enforcement workflow.
 *
 * Walks every `.proof` file in the project root (every package under
 * node_modules including scoped packages, plus the project root itself),
 * decodes each, and returns a unified
 * CID-keyed pool of all member mementos. All downstream stages do
 * hash lookups against this pool — no further file IO.
 *
 * Also indexes the bridge envelopes separately so the
 * enumerate-callsites stage can match Ctor names to bridges in O(1).
 *
 * Spec: protocol/specs/2026-04-30-proof-file-format.md (file walking +
 * trust-root verification)
 */

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { createHash } from "node:crypto";
import type { Stage } from "../types.js";
import { decodeProofEnvelope } from "../../proofEnvelope/index.js";
import { computeEnvelopeCid } from "../../claimEnvelope/cid.js";
import type { ClaimEnvelope, BridgeEvidence } from "../../claimEnvelope/types.js";

export const LOAD_ALL_PROOFS_CAPABILITY = "load-all-proofs";

export interface LoadAllProofsInput {
  projectRoot: string;
}

export interface LoadAllProofsOutput {
  /** CID → memento envelope. Every member of every .proof file. */
  mementoPool: Record<string, ClaimEnvelope>;
  /** sourceSymbol (IR name) → bridge envelope. Index for callsite enumeration. */
  bridgesBySymbol: Record<string, ClaimEnvelope>;
  /** Per-file errors encountered during the walk (failed trust root, etc.). */
  errors: Array<{ proofFile: string; reason: string }>;
}

export interface MakeLoadAllProofsStageDeps {
  producerVersion?: string;
}

export function makeLoadAllProofsStage(
  deps: MakeLoadAllProofsStageDeps = {},
): Stage<LoadAllProofsInput, LoadAllProofsOutput> {
  const producedBy = deps.producerVersion ?? "loadAllProofs@v1";

  return {
    name: "loadAllProofs",
    producedBy,

    serializeInput(input) {
      return { projectRoot: input.projectRoot };
    },

    serializeOutput(output) {
      return JSON.stringify(output);
    },

    deserializeOutput(witness) {
      return JSON.parse(witness) as LoadAllProofsOutput;
    },

    async run(input) {
      const mementoPool: Record<string, ClaimEnvelope> = {};
      const bridgesBySymbol: Record<string, ClaimEnvelope> = {};
      const errors: Array<{ proofFile: string; reason: string }> = [];

      const proofPaths = enumerateProofFiles(input.projectRoot);
      for (const proofPath of proofPaths) {
        let bytes: Buffer;
        try {
          bytes = readFileSync(proofPath);
        } catch (e) {
          errors.push({ proofFile: proofPath, reason: `read: ${(e as Error).message}` });
          continue;
        }
        // Trust-root check: filename CID matches bytes hash.
        const filename = proofPath.split("/").pop()!;
        const m = filename.match(/^([0-9a-f]+)\.proof$/);
        if (m) {
          const filenameCid = m[1]!;
          const derivedCid = createHash("sha256").update(bytes).digest("hex").slice(0, 32);
          if (derivedCid !== filenameCid) {
            errors.push({
              proofFile: proofPath,
              reason: `rule 1 (trust root): filename CID ${filenameCid} != content hash ${derivedCid}`,
            });
            continue;
          }
        }

        let catalog;
        try {
          catalog = decodeProofEnvelope(new Uint8Array(bytes));
        } catch (e) {
          errors.push({ proofFile: proofPath, reason: `decode: ${(e as Error).message}` });
          continue;
        }

        for (const [memberCid, memberBytes] of catalog.members) {
          let env: ClaimEnvelope;
          try {
            env = JSON.parse(Buffer.from(memberBytes).toString("utf8"));
          } catch (e) {
            errors.push({
              proofFile: proofPath,
              reason: `member ${memberCid.slice(0, 12)}…: parse: ${(e as Error).message}`,
            });
            continue;
          }
          // Verify the member's CID re-derives.
          const derived = computeEnvelopeCid(env);
          if (derived !== memberCid) {
            errors.push({
              proofFile: proofPath,
              reason: `member ${memberCid.slice(0, 12)}… bytes derive to ${derived}`,
            });
            continue;
          }
          mementoPool[memberCid] = env;
          if (env.evidence?.kind === "bridge") {
            const ev = env.evidence as BridgeEvidence;
            bridgesBySymbol[ev.body.sourceSymbol] = env;
          }
        }
      }

      return { mementoPool, bridgesBySymbol, errors };
    },
  };
}

/**
 * Enumerate every .proof file rooted at the project: the project root
 * itself plus each top-level package under node_modules (including
 * scoped packages one level deep). Excludes nested node_modules to
 * avoid combinatorial walks.
 */
function enumerateProofFiles(projectRoot: string): string[] {
  const out: string[] = [];

  // Project root's own *.proof files.
  if (existsSync(projectRoot)) {
    pushProofs(projectRoot, out);
  }

  const nodeModules = join(projectRoot, "node_modules");
  if (!existsSync(nodeModules)) return out;

  for (const entry of readdirSyncSafe(nodeModules)) {
    if (entry.startsWith(".")) continue;
    const entryPath = join(nodeModules, entry);
    let entryStat;
    try {
      entryStat = statSync(entryPath);
    } catch {
      continue;
    }
    if (!entryStat.isDirectory()) continue;

    if (entry.startsWith("@")) {
      for (const sub of readdirSyncSafe(entryPath)) {
        pushProofs(join(entryPath, sub), out);
      }
    } else {
      pushProofs(entryPath, out);
    }
  }
  return out;
}

function pushProofs(dir: string, out: string[]): void {
  for (const f of readdirSyncSafe(dir)) {
    if (f.endsWith(".proof")) out.push(join(dir, f));
  }
}

function readdirSyncSafe(p: string): string[] {
  try {
    return readdirSync(p);
  } catch {
    return [];
  }
}
